File: Leetcode_Hot_100/code/cpp/03_longest-substring-without-repeating-characters.cpp
#include <string>
#include <algorithm>

class Solution {
public:
    int lengthOfLongestSubstring(const std::string& s)
    {
        int left = 0, right = 0;
        int maxLength = 0;
        bool flag[128] = { false };
        while(right < s.size())
        {
            if (!flag[s[right]]) // 右指针滑动
            {
                flag[s[right]] = true;
                right++;
            }
            else
            {
                maxLength = std::max(maxLength, right - left);
                if (s[left] != s[right])
                {
                    flag[s[left]] = false;
                    left++;
                }
                else
                {
                    left++;
                    right++;
                }
            }
        }
        return std::max(maxLength, right - left);
    }
};

File: Leetcode_Hot_100/code/golang/03_longest-substring-without-repeating-characters.go
package main

func lengthOfLongestSubstring(s string) int {
	left, right, maxLength := 0, 0, 0
	flag := [128]bool{}
	for right < len(s) {
		if !flag[s[right]] {
			flag[s[right]] = true
			right++
		} else {
			if maxLength < (right - left) {
				maxLength = right - left
			}
			if s[left] != s[right] {
				flag[s[left]] = false
				left++
			} else {
				right++
				left++
			}
		}
	}
	if maxLength < (right - left) {
		return right - left
	} else {
		return maxLength
	}
}

func main() {

}
